models/db/engine.go

```diff
@@ -293,3 +293,25 @@ func GetMaxID(beanOrTableName interface{}) (maxID int64, err error) {
 	_, err = x.Select("MAX(id)").Table(beanOrTableName).Get(&maxID)
 	return
 }
+
+// QuoteString quotes the string for SQLs. In most cases, SQL should use the ORM builder.
+// This function could only be used in rare cases when there is no other choices.
+// It panics if the database is not supported to avoid corrupted data
+func QuoteString(ctx context.Context, s string) string {
+	e, ok := GetEngine(ctx).(*xorm.Engine)
+	if !ok {
+		panic("can not get the real database engine")
+	}
+
+	switch e.Dialect().URI().DBType {
+	case schemas.MYSQL:
+		s = strings.ReplaceAll(s, "\\", "\\\\")
+		s = strings.ReplaceAll(s, "'", "''")
```

Comment:
```suggestion
		rv := []rune(s)
		buf := make([]rune, len(rv)*2)
		pos := 0
		for i := 0; i < len(rv); i++ {
			c := rv[i]
			switch c {
			case '\x00':
				buf[pos] = '\\'
				buf[pos+1] = '0'
				pos += 2
			case '\n':
				buf[pos] = '\\'
				buf[pos+1] = 'n'
				pos += 2
			case '\r':
				buf[pos] = '\\'
				buf[pos+1] = 'r'
				pos += 2
			case '\x1a':
				buf[pos] = '\\'
				buf[pos+1] = 'Z'
				pos += 2
			case '\'':
				buf[pos] = '\\'
				buf[pos+1] = '\''
				pos += 2
			case '"':
				buf[pos] = '\\'
				buf[pos+1] = '"'
				pos += 2
			case '\\':
				buf[pos] = '\\'
				buf[pos+1] = '\\'
				pos += 2
			default:
				buf[pos] = c
				pos++
			}
		}

		return "'" + string(buf[:pos]) + "'"
```

Reply:
Are you sure we have to do this? AFAIU you only need the escaping on connections that would otherwise fail.

Xorm itself doesn't do this amount of escaping in its INSERTs

Reply:
(Although I do remember there are several problems with `\0` but I think these are in general disallowed within SQL strings)

Reply:
https://github.com/go-sql-driver/mysql/blob/ad9fa14acdcf7d0533e7fbe58728f3d216213ade/utils.go#L692

Reply:
Xorm doesn't use the variable binding within INSERT...

Reply:
It's not a must. PHP used [addslashes](https://www.php.net/manual/en/function.addslashes.php) in history for SQL quoting (although it's not the recommended way anymore), it also only encodes a few characters.

MySQL & MariaDB can always handle the ASCII chars between 0x01~0x1F. It's not necessary to make it too complex. They won't break such convention.

And the current code matches XORM [convertString](https://gitea.com/xorm/xorm/src/commit/26d291bbc311274388045768386f015a06b3be10/internal/statements/statement_args.go#L56).

So, it should be kept as it was.

And we already have testcases for all ASCII characters.

Reply:
I added more comments to this function, hopefully nobody will be hurt in the future 😂

Reply:
I added more comments to this function, hopefully nobody will be hurt in the future 😂

Reply:
See comments in MySQL code on why it's needed in some cases: https://github.com/mysql/mysql-server/blob/mysql-5.7.5/mysys/charset.c#L883

Reply:
I know that's for escaping. But for unescaping, the unesacping algorithm can accept the simple escaping.

Reply:
If you believe there is a bad case, you could propose a test case and make the code cover it.

Reply:
> Xorm doesn't use the variable binding within INSERT...

Are you sure?